Dockerfile

```diff
@@ -20,75 +20,67 @@ ARG COMMIT_SHA
 ENV COMMIT_SHA=${COMMIT_SHA:-}
 ENV REACT_APP_BUILD_COMMIT_SHA=${COMMIT_SHA:-}
 
-RUN mkdir -p /app
-WORKDIR /app
-
 # Download dependencies
-COPY ui/package*.json /app/
-RUN npm ci --omit=dev --ignore-scripts
-
-# Build frontend
 COPY ui /app
-RUN npm run build -- --base=$UI_PUBLIC_URL
+WORKDIR /app
+RUN npm ci --omit=dev --ignore-scripts \
+  # Build frontend
+  && npm run build -- --base=$UI_PUBLIC_URL
 
 ### Build API ###
 FROM --platform=${BUILDPLATFORM:-linux/amd64} debian:bookworm AS api
 ARG TARGETPLATFORM
 
 COPY docker/install_build_dependencies.sh /tmp/
-RUN chmod +x /tmp/install_build_dependencies.sh && /tmp/install_build_dependencies.sh
-
 COPY docker/go_wrapper.sh /go/bin/go
-RUN chmod +x /go/bin/go
+COPY api /app
+WORKDIR /app
+
 ENV GOPATH="/go"
 ENV PATH="${GOPATH}/bin:${PATH}"
-
 ENV CGO_ENABLED 1
 
-RUN go env
-
-RUN mkdir -p /app
-WORKDIR /app
-
 # Download dependencies
-COPY api/go.mod api/go.sum /app/
-RUN go mod download
-
-# Patch go-face
-RUN sed -i 's/-march=native//g' ${GOPATH}/pkg/mod/github.com/!kagami/go-face*/face.go
-
-# Build dependencies that use CGO
-RUN go install \
-  github.com/mattn/go-sqlite3 \
-  github.com/Kagami/go-face
-
-# Copy and build api source
-COPY api /app
-RUN go build -v -o photoview .
+RUN chmod +x /tmp/install_build_dependencies.sh \
+  && chmod +x /go/bin/go \
+  && /tmp/install_build_dependencies.sh \
+  && go env \
+  && go mod download \
+  # Patch go-face
+  && sed -i 's/-march=native//g' ${GOPATH}/pkg/mod/github.com/!kagami/go-face*/face.go \
+  # Build dependencies that use CGO
+  && go install \
+    github.com/mattn/go-sqlite3 \
+    github.com/Kagami/go-face \
+  # Build api source
+  && go build -v -o photoview .
 
 ### Copy api and ui to production environment ###
-FROM debian:bookworm
+FROM debian:bookworm-slim
 ARG TARGETPLATFORM
 WORKDIR /app
 
-COPY api/data /app/data
-
-RUN apt update \
+# Create a user to run Photoview server
+RUN groupadd -r photoview \
+  && useradd -r -g photoview photoview \
+  && mkdir -p /home/photoview \
+  && chown -R photoview:photoview /app /home/photoview \
```

Comment:
This can be greatly simplified:
1. `useradd` can take care of the user group (`-U`) as well as the home directory (`-m`, optionally with `-d /path/to/user/dir` if you want to set a custom directory)
2. The app dir shouldn't be chowned so this can be removed

```suggestion
RUN useradd -r -U -m photoview \
```

Reply:
One side-note on the `/app` ownership; the Dockerfile currently sets it as the WORKDIR, which in turn results in the cache location defaulting to `/app/media_cache` if not otherwise specified by the user by setting `PHOTOVIEW_MEDIA_CACHE`.  
To ensure a good out-of-the-box experience for that situation, the above suggestion could be expanded to prepare this default directory for the end-user, e.g.:
```suggestion
RUN useradd -r -U -m photoview \
  && install -dm755 -g photoview -o photoview /app/media_cache \
```

Reply:
> Why? This is the application folder, why the application user shouldn't own it?  

The application should be available for running by the (any) user but you don't want the user (or anything and everything getting its privileges) to just modify w/e it wants in those applications.  
Especially for internet facing services doing so would be a risk; the moment an attacker manages to break in, they'll immediately have the user's privileges and thus the capabilities to edit all the files that are served to end-users. When those files are instead owned by root and only executable for other users, this can not be done without also somehow managing a privilege escalation first.  
The executing user should only ever need to have execution permissions for the application. The only ownership it needs is over its own user directories.  

When you look at your own system, you should see very similar behaviour. Your (/usr)/(s)bin isn't owned by your user but by root. You use a package manager, run with root permissions, to install and update your applications but you then execute the individual applications using your own user's privileges. It does not have the capabilities to change those binaries but any runtime data is written to your user's folders (e.g. the XDG directories for data, config and cache, dot-files in $HOME, etc.)

Reply:
Fwiw, an alternative for the default media cache situation could also be to just set it to the user dir as part of the Dockerfile, e.g. `ENV PHOTOVIEW_MEDIA_CACHE=/home/photoview/media_cache`, it doesn't need to be prepared in that case as the home is already set up during `useradd`.

Reply:
> One side-note on the `/app` ownership; the Dockerfile currently sets it as the WORKDIR, which in turn results in the cache location defaulting to `/app/media_cache` if not otherwise specified by the user by setting `PHOTOVIEW_MEDIA_CACHE`. To ensure a good out-of-the-box experience for that situation, the above suggestion could be expanded to prepare this default directory for the end-user, e.g.:
> `&& install -dm755 -g photoview -o photoview /app/media_cache \`

Why do we need to drop chowning the root application folder, but then create in advance and chown the nested folder for the cache? what is the real benefit of such self-limiting actions?

I didn't even test this, but if this will work and by now Photoview doesn't require the ownership of the /app folder, it might be changed in the future and this limitation is not so obvious - which creates a potential headache for future debugging and maintenance (similar to the issue with the Darktable, failed to run without having write access to the user's home folder)

Reply:
> Why do we need to drop chowning the root application folder, but then create in advance and chown the nested folder for the cache?  

The only reason this is needed is because the Dockerfile sets the WORKDIR to /app. There's in total 2 options you can take to prevent this from being needed in the first place:

1. Set the WORKDIR to the user's $HOME
2. Set the PHOTOVIEW_MEDIA_CACHE (see also https://github.com/photoview/photoview/pull/863#discussion_r1576304311 )

One of these would have my preference over the other case anyway but I wasn't sure what the expectations around WORKDIR and PHOTOVIEW_MEDIA_CACHE are so I went with the default for that to base my first suggestion on.


> what is the real benefit of such self-limiting actions?  

There is no self-limiting action here. There's only fixing inconsistencies, preventing bad practices and resolving security problems with the proposed solution.  


> it might be changed in the future and this limitation is not so obvious  

That should never be made to happen. Binary and runtime/config dirs should _always_ be kept separate. The application should never, ever, write in it's binary directory.  


> which creates a potential headache for future debugging and maintenance (similar to the issue with the Darktable, failed to run without having write access to the user's home folder)  

Which had nothing to do with the application folder but everything with the user directories, exactly that which this PR should resolve and straighten out in the first place doesn't it??